beacon-chain/blockchain/service.go

```diff
@@ -234,16 +253,99 @@ func (c *ChainService) blockProcessing(processedBlock chan<- *types.Block) {
 	}
 }
 
+func (c *ChainService) blockProcessingNew(processedBlock chan<- *types.Block) {
+	subBlock := c.incomingBlockFeed.Subscribe(c.incomingBlockChan)
+	defer subBlock.Unsubscribe()
+	for {
+		select {
+		case <-c.ctx.Done():
+			log.Debug("Chain service context closed, exiting goroutine")
+			return
+
+		// Listen for a newly received incoming block from the feed. Blocks
+		// can be received either from the sync service, the RPC service,
+		// or via p2p.
+		case block := <-c.incomingBlockChan:
+			if c.currentSlot < block.SlotNumber() {
+				c.unProcessedBlocks[block.SlotNumber()] = block
+				continue
+			}
+
+			if c.currentSlot > block.SlotNumber() {
+				log.Debugf(
+					"Block slot number is lower than the currently processed slot in the beacon state %d",
+					block.SlotNumber())
+				continue
+			}
+
+			if err := c.processBlock(block); err != nil {
+				log.Error(err)
+				processedBlock <- nil
+				continue
+			}
+
+			// Push the block to trigger the fork choice rule.
+			processedBlock <- block
+		}
+	}
+}
+
+// This assigns the current slot of the
+// beacon node, functioning as the beacon node's
+// local clock.
+func (c *ChainService) slotTracker(slotTicker <-chan uint64, tickerDone func()) {
+
+	defer tickerDone()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			log.Debug("Chain service context closed, exiting goroutine")
+			return
+		case slot := <-slotTicker:
+			c.currentSlot = slot
+
+			c.checkForSkippedSlots()
+
+			beaconState, err := c.beaconDB.GetState()
+			if err != nil {
+				log.Debugf("Unable to retrieve beacon state %v", err)
+				continue
+			}
+
+			// Setting per slot counters
+
+			// Updating beacon state slot
+			beaconState.SetSlot(slot)
+			vreg := beaconState.ValidatorRegistry()
+
+			proposerIndex, err := v.GetBeaconProposerIndex(beaconState.Proto(), beaconState.Slot())
+			if err != nil {
+				log.Debugf("Unable to retrieve proposer index %v", err)
+				continue
+			}
+
+			// Updating proposer randao layers, might be a bug, for reference:
+			// https://github.com/ethereum/eth2.0-specs/issues/319
+			vreg[proposerIndex].RandaoLayers++
+			beaconState.SetValidatorRegistry(vreg)
+
+			if err := c.beaconDB.SaveState(beaconState); err != nil {
+				log.Debugf("Unable to save beacon state to db: %v", err)
+			}
+
```

Comment:
Should we be updating recent block hashes here as well?

Reply:
+1

Reply:
I think you did that already `newState.SetLatestBlockHashes(newhashes)`